data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostAttestationsV2.java

```diff
@@ -95,22 +106,46 @@ private static EndpointMetadata createMetadata(
 
     builder.withType(
         value -> attestationSchemaPredicate.test(value, SpecMilestone.PHASE0),
-        schemaDefinitionCache
-            .getSchemaDefinition(SpecMilestone.PHASE0)
-            .getAttestationSchema()
-            .getJsonTypeDefinition());
+        getPhaseOAttestationSchema(schemaDefinitionCache).getJsonTypeDefinition());
     builder.withType(
         value -> attestationSchemaPredicate.test(value, SpecMilestone.ELECTRA),
-        SchemaDefinitionsElectra.required(
-                schemaDefinitionCache.getSchemaDefinition(SpecMilestone.ELECTRA))
-            .getSingleAttestationSchema()
-            .getJsonTypeDefinition());
+        getElectraAttestationSchema(schemaDefinitionCache).getJsonTypeDefinition());
     final SerializableOneOfTypeDefinition<Attestation> attestationSchemaDefinition =
         builder.build();
 
     final OneOfArrayJsonRequestContentTypeDefinition.BodyTypeSelector<Attestation>
         attestationBodySelector =
-            context -> headerBasedSelector(context.getHeaders(), schemaDefinitionCache);
+            context -> {
+              final Map<String, String> headers = context.getHeaders();
+              if (!headers.containsKey(HEADER_CONSENSUS_VERSION)) {
+                throw new BadRequestException(
+                    String.format(
+                        "Missing required header value for (%s)", HEADER_CONSENSUS_VERSION));
+              }
```

Comment:
are we moving away from headerBasedSelector just for that error? we're probably better off just using the selector, we still need the typedef...